docs/proposals/006-scheduler/README.md

```diff
@@ -0,0 +1,150 @@
+
+1. Fast iteration with the ML ecosystem, namely other languages
+2. Benefit from existing informers without having multiple implementations
+3. Acceptable speed of scheduling for 10-1000 QPS systems
+
+### Design
+
+We expect the following challenges to be addressed by the reference scheduler design:
+
+1. Understand the cost of an incoming request and its impact on the target model server before placing it
+2. Track the cost of previously issued requests to avoid overloading servers
+3. Integrate future cost features such as prefix cache routing into a holistic cost model
+4. Support heterogenous model server capabilities in terms of capacity, latency, memory, and features
+
+#### Reference Scheduler
+
+The reference scheduler will be a monolithic Golang scheduler that is expected to run cooperatively with other instances of the scheduler with the same configuration or with appropriate 1 version/config skew.
+
+The reference scheduler receives a list of **candidate endpoints** from the EPP and is responsible for selecting a match.
+
+The reference scheduler is **informed** about the current state of model servers via **informers**, of which the current informer is a fast-polling loop retrieving model server metrics via the [model server protocol](../003-model-server-protocol/).
+
+The reference scheduler is configured with a series of **predicates** that **filter** candidate endpoints, removing impossible matches.  If no matches or only one match is feasible, that endpoint is selected. If multiple matches are made, the scheduler will consult a list of configured **scorers** to **score** the matches into a **prioritized** list of endpoints, and then **sample** from that list.
+
+Once an endpoint is selected, the endpoint is **assumed** to be running that request until the EPP observes the termination of that request (most common) OR an informer invalidates the execution of those requests.  The scheduler must integrate the impact of assumed load to with informer state, especially when traffic spikes.
+
+Given that we anticipate a significant amount of future work to integrate heterogenous hardware (different generations / topologies) and heterogeous server roles (prefill-heavy, prefill/decode split, latency objectives), we expect that there will be an **assignment** informer that partitions the candidate endpoints over multiple dimensions for the scheduler.  This will decouple the scheduling algorithm from the process of determining the capacity and suitability of different model servers to different dimensions of request cost.
+
+#### Replacement Scheduler
+
+The replacement scheduler will be a low-latency mechanism for out-of-process execution of the core endpoint selection option.  The replacement scheduler will accept one or more requests to schedule, a list of endpoints, and optionally the associated informer state for those endpoints. The replacement scheduler will return one or zero endpoints per request.
```

Comment:
Why does this look like a batch scheduler?

Reply:
1k qps will require either streaming or batching, fair point we don’t want to constrain that design yet